histogram/src/sparse.rs

```diff
@@ -0,0 +1,304 @@
+        let mut seen: usize = 0;
+        for (idx, count) in self.index.iter().zip(self.count.iter()) {
+            seen += *count as usize;
+            if seen >= search {
+                return Ok(Bucket {
+                    count: *count,
+                    range: self.config.index_to_range(*idx),
+                });
+            }
+        }
+
+        // should be unreachable
+        Err(Error::Unreachable)
+    }
+
+    /// Returns a new histogram with a reduced grouping power. The specified
+    /// reduction factor should be 0 < factor < existing grouping power.
+    ///
+    /// This works by iterating over every bucket in the existing histogram
+    /// and inserting the contained values into the new histogram. While we
+    /// do not know the exact values of the data points (only that they lie
+    /// within the bucket's range), it does not matter since the bucket is
+    /// not split during downsampling and any value can be used.
+    pub fn downsample(&self, factor: u8) -> Result<SparseHistogram, Error> {
+        let grouping_power = self.config.grouping_power();
+
+        if factor == 0 || grouping_power <= factor {
+            return Err(Error::MaxPowerTooLow);
+        }
+
+        let config = Config::new(grouping_power - factor, self.config.max_value_power())?;
+        let mut histogram = SparseHistogram {
+            config,
+            index: Vec::new(),
+            count: Vec::new(),
+        };
+
+        // Multiple buckets in the old histogram will map to the same bucket
+        // in the new histogram, so we have to aggregate bucket values from the
+        // old histogram before inserting a bucket into the new downsampled
+        // histogram. However, mappings between the histograms monotonically
+        // increase, so once a bucket in the old histogram maps to a higher
+        // bucket in the new histogram than is currently being aggregated,
+        // the bucket can be sealed and inserted into the new histogram.
+        let mut aggregating_idx: usize = 0;
+        let mut aggregating_count: u64 = 0;
+        for (idx, n) in self.index.iter().zip(self.count.iter()) {
+            let new_idx = config.value_to_index(self.config.index_to_lower_bound(*idx))?;
+
+            // If it maps to the currently aggregating bucket, merge counts
+            if new_idx == aggregating_idx {
+                aggregating_count += n;
+                continue;
+            }
+
+            // Does not map to the aggregating bucket, so seal and store that bucket
+            histogram.add_bucket(aggregating_idx, aggregating_count);
+
+            // Start tracking this bucket as the current aggregating bucket
+            aggregating_idx = new_idx;
+            aggregating_count = *n;
+        }
+
+        // Add the final aggregated bucket
+        histogram.add_bucket(aggregating_idx, aggregating_count);
+
+        Ok(histogram)
+    }
+}
+
+impl From<&Snapshot> for SparseHistogram {
```

Comment:
Should we `impl From<&Histogram>` ? Snapshot -> SparseHistogram causes loss of the start and end timestamps. I'm wondering if this is really what we want to do here.

Reply:
Chatted offline: there are conversions from `Histogram` and `Snapshot` but not from `AtomicHistogram` which remains append-only.